### ISSUE TITLE
Local (folder relative) Node packages will be ignored by `tsc` when typechecking

### ISSUE THREAD

**Issue by fatso83** [microsoft/TypeScript#56576]
### 🔎 Search Terms

"symbolic links", symlinks, "local packages", ".."

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about symbolic links
- I was unable to test this on versions prior to 4.3, because the behavior was present  down to 4.3, but below 4.2 the dependencies in my project was using too new Typescript syntax

### ⏯ Playground Link

https://github.com/fatso83/retry-dynamic-import/tree/typescript-symlink-issue

### 💻 Code

 npx tsc --noEmit typescript-moduleresolution-default/src/main.ts

### 🙁 Actual behavior

No errors are produced. This happens because `tsc` is not actually inspecting any typescript definitions when it encounters a symbolic link in `node_modules/@fatso83/retry-dynamic-import`. This is problematic when I try to check definitions before publishing, by consuming locally produced types in a consuming package.

### 🙂 Expected behavior

It should fail, as the current types have multiple errors in them relating to augmenting of modules.

### Additional information about the issue

This is easily verifiable in a minute or two by following the [README](https://github.com/fatso83/retry-dynamic-import/tree/typescript-symlink-issue). Just install deps as normal, run tests (using local symlinks) and then after removing the symlink and copying the files, the same tests will fail with typechecking errors.

I have verified the behavior is present in TSC for 4.3-5.0.4

**Comment by andrewbranch** [comment]
Thanks for the detailed repro instructions.

> No errors are produced. This happens because `tsc` is not actually inspecting any typescript definitions

This is not true. You can verify with `tsc --listFilesOnly`. The reason you’re not getting errors after deleting the type definitions entirely is you’re not compiling with `strict`. Missing type definitions for packages are only errors under `--noImplicitAny`.

The reason you’re not consistently getting errors when `pkg` is symlinked is that this module specifier

```ts
declare module "@fatso83/retry-dynamic-import/react-lazy" {
  // ...
}
```

is not resolvable from the file’s realpath location, because there is no `node_modules/@fatso83/...` directory in scope of `pkg`. Declaration files are allowed to declare module augmentations that fail to resolve so that one package can conditionally augment another if present, without causing errors if the other package isn’t installed. So when you have a compilation where that module augmentation doesn’t error, it’s because it can’t find anything to augment. Once module resolution succeeds, the augmentation becomes invalid due to duplicate identifiers or whatever.

If you want to make a minimal change to your setup to get this working, you’ll need to move the `node_modules/@fatso83/retry-dynamic-import` symlink to the repo’s top-level node_modules, as opposed to `test-builds/node_modules`. However, I strongly recommend using something like pnpm workspaces to handle monorepo development in a more conventional way.
